lib/charms/opensearch/v0/opensearch_nodes_exclusions.py

```diff
@@ -50,17 +47,22 @@ def add_current(self, restart: bool = False) -> None:
             logger.error(f"Failed to add voting exclusion: {self._node.name}.")
 
         if not restart:
-            if self._node.is_data() and not self._add_allocations():
+            if self._node.is_data() and not self.add_allocations():
                 logger.error(f"Failed to add shard allocation exclusion: {self._node.name}.")
 
     def delete_current(self) -> None:
         """Delete Voting and alloc exclusions."""
-        if (
-            self._node.is_cm_eligible() or self._node.is_voting_only()
-        ) and not self._delete_voting():
-            self._charm.peers_data.put(self._scope, VOTING_TO_DELETE, True)
-
-        if self._node.is_data() and not self._delete_allocations():
+        if self._node.is_cm_eligible() or self._node.is_voting_only():
+            if self._delete_voting():
+                self._charm.peers_data.put(
+                    self._scope,
+                    VOTING_TO_DELETE,
+                    ",".join(self._fetch_voting_exclusions()),
+                )
```

Comment:
Can you explain why we fetch and store all exclusions? my understanding was that we only keep track on the relations data of the ones that failed to be deleted (similar to the logic of allocations below)

Reply:
Voting exclusions sits on the critical path of the 3<->2<->1 nodes scaling. It does not apply to all the cases.

Now, on these cases, we **must** apply them.

However, in this specific case, I **must** apply the voting exclusion before, say, scaling 3->2, or clean it up before 1->2 scaling. The reason is because once we reach 2x nodes, then we cannot guarantee the API will work forever, as it may eventually try another election and we lose the known leader. 

So, indeed, keeping track of the exclusions in a peer relation is not needed. We can fetch the results anyways. I kept that only to have some similarity with allocations.

Reply:
Giving it some extra thought, I will do the following: use it to keep track of our own voting exclusions only instead of a `fetch`.

Reply:
@Mehdi-Bendriss I created another logic:
* `CURRENT_EXCLUSIONS`: a peer databag list that tracks the exclusions defined by the charm
* At `_add_voting`: we add one unit to `CURRENT_EXCLUSIONS` and a new voting exclusion
* At `_cleanup_voting`: (1) **potential nodes to remove**: check the list of nodes present in the full voting exclusions and not present in our deployment; (2) iterate over `CURRENT_EXCLUSIONS`: if a node is present in this list AND in the **potential nodes to remove**, then we should take this node out of the voting exclusions as it is occupying a vote and it is actually gone
* At `_delete_voting`: (1) run the _cleanup_voting; (2) remove the node itself from that list as well; and (3) remove all votes and readd the remaining ones to the list

This way, we make sure we can preserve any user-defined changes but still keep the list clean.

Some questions:
1) We should use the peer-cluster relation as well to publish CURRENT_EXCLUSIONS, right? Otherwise, if we have a failover, then the list is now present in the peer databag of the `failover cluster`
2) We have a hard limit of voting count (10 in general), there is a risk we will hit this limit if we let the user also set voting and we respect these values, what should we do in this case?

Reply:
I think we could potentially simplify it and build on top of what we already have by only making use of `ALLOCS_TO_DELETE` and `VOTING_TO_DELETE` which we update whenever we have a failed deletion, and recurrent cleanup routine attempts to wipe the content of these no?